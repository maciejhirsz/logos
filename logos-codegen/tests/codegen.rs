use std::{error::Error, io, path::PathBuf};

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let mut fixture_dir = PathBuf::new();
    fixture_dir.push(env!("CARGO_MANIFEST_DIR"));
    fixture_dir.push("tests");
    fixture_dir.push("data");
    fixture_dir.push(fixture);

    let input = fixture_dir.join("input.rs");
    fixture_dir.push("output.rs");
    let output_file = fixture_dir;

    let input = std::fs::read_to_string(input)?;
    let output = std::fs::read_to_string(&output_file)?;

    let generated = logos_codegen::generate(input.parse()?);
    let generated = generated.to_string();

    if std::env::var("BLESS_CODEGEN").is_ok_and(|value| value == "1") {
        std::fs::write(&output_file, &generated)?;
        return Ok(());
    }

    assert_eq!(generated, output, "Codegen test failed: `{fixture}`, run tests again with env var `BLESS_CODEGEN=1` to bless these changes");

    Ok(())
}
