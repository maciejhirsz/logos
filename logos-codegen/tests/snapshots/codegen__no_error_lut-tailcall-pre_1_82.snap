---
source: logos-codegen/tests/codegen.rs
expression: formatted
---
impl<'s> ::logos::Logos<'s> for Token {
    type Error = ();
    type Extras = ();
    type Source = [u8];
    fn lex(
        lex: &mut ::logos::Lexer<'s, Self>,
    ) -> std::option::Option<std::result::Result<Self, Self::Error>> {
        use ::logos::internal::{
            LexerInternal, CallbackRetVal, CallbackResult, SkipRetVal, SkipResult,
        };
        use std::result::Result as _Result;
        use std::option::Option as _Option;
        use ::logos::Logos;
        type _Lexer<'s> = ::logos::Lexer<'s, Token>;
        #[inline]
        fn _make_error<'s>(lex: &mut _Lexer<'s>) -> <Token as Logos<'s>>::Error {
            <Token as Logos<'s>>::Error::default()
        }
        fn state1024_ctx2<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset - 1);
            match lex.read::<u8>(offset) {
                other => {
                    if other.is_none() {}
                    return Some(Ok(Token::Any));
                }
            }
        }
        fn state1536_ctx2<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset - 1);
            match lex.read::<u8>(offset) {
                Some(128u8..=191u8) => {
                    offset += 1;
                    return state7680_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {}
                    return Some(Ok(Token::Any));
                }
            }
        }
        fn state2048_ctx1<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset - 1);
            match lex.read::<u8>(offset) {
                other => {
                    if other.is_none() {}
                    return Some(Ok(Token::AnyUnicode));
                }
            }
        }
        fn state2560_ctx2<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset - 1);
            match lex.read::<u8>(offset) {
                Some(128u8..=191u8) => {
                    offset += 1;
                    return state8192_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {}
                    return Some(Ok(Token::Any));
                }
            }
        }
        fn state3072_ctx2<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset - 1);
            match lex.read::<u8>(offset) {
                Some(0u8..=255u8) => {
                    offset += 1;
                    return state2048_ctx1(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state2048_ctx1(lex, offset);
                    }
                    return Some(Ok(Token::Any));
                }
            }
        }
        fn state3584_ctx0<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset - 1);
            match lex.read::<u8>(offset) {
                other => {
                    if other.is_none() {}
                    return Some(Ok(Token::Newline));
                }
            }
        }
        fn state4096_ctx1<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset - 1);
            match lex.read::<u8>(offset) {
                other => {
                    if other.is_none() {}
                    return Some(Ok(Token::AnyUnicode));
                }
            }
        }
        fn state4608<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {
                enum LogosNextState {
                    state5120,
                    state5632,
                    state6144,
                    state6656,
                    state7168,
                    state8704,
                    state9216,
                    state9728,
                    state10240,
                    state10752,
                }
                const TABLE: [_Option<LogosNextState>; 256] = [
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state8704),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state7168),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state9728),
                    Some(LogosNextState::state10240),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state6656),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state10752),
                    Some(LogosNextState::state5120),
                    Some(LogosNextState::state5632),
                    Some(LogosNextState::state5632),
                    Some(LogosNextState::state5632),
                    Some(LogosNextState::state6144),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                    Some(LogosNextState::state9216),
                ];
                match TABLE[byte as usize] {
                    Some(LogosNextState::state5120) => {
                        offset += 1;
                        return state5120(lex, offset);
                    }
                    Some(LogosNextState::state5632) => {
                        offset += 1;
                        return state5632(lex, offset);
                    }
                    Some(LogosNextState::state6144) => {
                        offset += 1;
                        return state6144(lex, offset);
                    }
                    Some(LogosNextState::state6656) => {
                        offset += 1;
                        return state6656(lex, offset);
                    }
                    Some(LogosNextState::state7168) => {
                        offset += 1;
                        return state7168(lex, offset);
                    }
                    Some(LogosNextState::state8704) => {
                        offset += 1;
                        return state8704(lex, offset);
                    }
                    Some(LogosNextState::state9216) => {
                        offset += 1;
                        return state9216(lex, offset);
                    }
                    Some(LogosNextState::state9728) => {
                        offset += 1;
                        return state9728(lex, offset);
                    }
                    Some(LogosNextState::state10240) => {
                        offset += 1;
                        return state10240(lex, offset);
                    }
                    Some(LogosNextState::state10752) => {
                        offset += 1;
                        return state10752(lex, offset);
                    }
                    None => {}
                }
            } else {
                if lex.offset() == offset {
                    return None;
                }
            }
            lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
            return Some(Err(_make_error(lex)));
        }
        fn state5120<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=143u8) | Some(192u8..=255u8) => {
                    offset += 1;
                    return state1024_ctx2(lex, offset);
                }
                Some(144u8..=191u8) => {
                    offset += 1;
                    return state1536_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state1024_ctx2(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        fn state5632<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=127u8) | Some(192u8..=255u8) => {
                    offset += 1;
                    return state1024_ctx2(lex, offset);
                }
                Some(128u8..=191u8) => {
                    offset += 1;
                    return state1536_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state1024_ctx2(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        fn state6144<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=127u8) | Some(144u8..=255u8) => {
                    offset += 1;
                    return state1024_ctx2(lex, offset);
                }
                Some(128u8..=143u8) => {
                    offset += 1;
                    return state1536_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state1024_ctx2(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        fn state6656<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=127u8) | Some(160u8..=255u8) => {
                    offset += 1;
                    return state1024_ctx2(lex, offset);
                }
                Some(128u8..=159u8) => {
                    offset += 1;
                    return state2560_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state1024_ctx2(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        fn state7168<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=255u8) => {
                    offset += 1;
                    return state4096_ctx1(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state4096_ctx1(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        fn state7680_ctx2<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(128u8..=191u8) => {
                    offset += 1;
                    return state8192_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {}
                    return Some(Ok(Token::Any));
                }
            }
        }
        fn state8192_ctx2<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=255u8) => {
                    offset += 1;
                    return state2048_ctx1(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state2048_ctx1(lex, offset);
                    }
                    return Some(Ok(Token::Any));
                }
            }
        }
        fn state8704<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=255u8) => {
                    offset += 1;
                    return state3584_ctx0(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state3584_ctx0(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        fn state9216<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=255u8) => {
                    offset += 1;
                    return state1024_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state1024_ctx2(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        fn state9728<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=127u8) | Some(192u8..=255u8) => {
                    offset += 1;
                    return state1024_ctx2(lex, offset);
                }
                Some(128u8..=191u8) => {
                    offset += 1;
                    return state3072_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state1024_ctx2(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        fn state10240<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=159u8) | Some(192u8..=255u8) => {
                    offset += 1;
                    return state1024_ctx2(lex, offset);
                }
                Some(160u8..=191u8) => {
                    offset += 1;
                    return state2560_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state1024_ctx2(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        fn state10752<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            match lex.read::<u8>(offset) {
                Some(0u8..=127u8) | Some(192u8..=255u8) => {
                    offset += 1;
                    return state1024_ctx2(lex, offset);
                }
                Some(128u8..=191u8) => {
                    offset += 1;
                    return state2560_ctx2(lex, offset);
                }
                other => {
                    if other.is_none() {
                        offset += 1;
                        return state1024_ctx2(lex, offset);
                    }
                    lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
                    return Some(Err(_make_error(lex)));
                }
            }
        }
        state4608(lex, lex.offset())
    }
}
